waku/v2/node/peer_manager/peer_manager.nim

```diff
@@ -284,3 +287,38 @@ proc connectToNodes*(pm: PeerManager,
   # This issue was known to Dmitiry on nim-libp2p and may be resolvable
   # later.
   await sleepAsync(chronos.seconds(5))
+
+# Ensures a healthy amount of connected relay peers
+proc relayConnectivityLoop*(pm: PeerManager) {.async.} =
+  let defaultInterval = chronos.seconds(30)
+
+  while true:
+
+    let maxConnections = pm.switch.connManager.inSema.size
+    let numInPeers = pm.switch.connectedPeers(lpstream.Direction.In).len
+    let numOutPeers = pm.switch.connectedPeers(lpstream.Direction.Out).len
+    let numConPeers = numInPeers + numOutPeers
+
+    # TODO: Enforce a given in/out peers ratio
+
+    # Leave some room for service peers
+    if numConPeers >= (maxConnections - 5):
+      await sleepAsync(defaultInterval)
+      continue
+
+    # TODO: Respect backoff before attempting to connect a relay peer
+    var disconnectedPeers = pm.peerStore.getDisconnectedPeers().mapIt(RemotePeerInfo.init(it.peerId, it.addrs))
+    shuffle(disconnectedPeers)
+
+    # limit the amount of paralel dials
+    let maxParalelDials = 10
+    let numPeersToConnect = min(min(maxConnections - numConPeers, disconnectedPeers.len), maxParalelDials)
+
+    info "Relay connectivity loop",
+      connectedPeers = numConPeers,
+      targetConnectedPeers = maxConnections,
+      availableDisconnectedPeers = disconnectedPeers.len
+
+    await pm.connectToNodes(disconnectedPeers[0..<numPeersToConnect], WakuRelayCodec)
```

Comment:
This is not a good strategy imo, if every peers try to dial peers until it reaches maxConnections, you will naturally have more outgoing connections than incoming connections, and that cannot work at the network scale (every outgoing connection is someone else's incoming connection)

Ideally, every node leaves `> maxConnections / 2` open for incoming connections (also need to account for peers that are "outbound only" and are just using slots without giving any)

What is done in nimbus is that we only dial when we have to (while we are unhealthy), and once we are healthy, we leave all the remaining slots for incoming connections. This makes nimbus very nice to the network (ie on our fleet, we generally have ~30 outgoing connections and >200 incoming). But if we have to, we'll kick every one of this incoming connection to get healthy (in case of attack or whatev)

Reply:
great input thanks.

with this PR i dont intend to add this in/out balance (see above "Enforce a given in/out peers ratio") but just decouple this logic from discovery without extra functionality.

but plan a follow up PR with this. Mind pointing to where this is done in nimbus? Can't immediatly see looking into `connectWorker` and `runDiscoveryLoop`

Reply:
https://github.com/status-im/nimbus-eth2/blob/f6a5a5b1da51863b2db9f812e5c682ca4eaa40ab/beacon_chain/networking/eth2_network.nim#L1562-L1563

It will not connect to anyone if we are already healthy

Not sure what the current behavior is in nwaku, but if it is similar to what is done here, it should be fixed ASAP, otherwise, as soon as there will be more than `maxConnections` peers in the network, the bootstrap nodes will be full of useless outgoing connections, and new peers won't be able to connect to them

Reply:
> Not sure what the current behavior is in nwaku, but if it is similar to what is done here, it should be fixed ASAP, otherwise, as soon as there will be more than maxConnections peers in the network, the bootstrap nodes will be full of useless outgoing connections, and new peers won't be able to connect to them

yep, this is the current behaviour. luckily we have more maxAmountOfConnections that nodes exists in any network. but yes, looking into it to fix it asap :)

Reply:
to be addressed here:
https://github.com/waku-org/nwaku/issues/1495